# Feature selection version 1

**Motivation:**  
This script creates **Feature version 1**.

**Version 1 definition:**
- Select all titles that appeared in the titulary of more than 10 viziers.  
    - That is, select titles satisfying: `count(title | vizier) > 10`.

### loads

In [1]:
import numpy as np
import pandas as pd
import set_path
import supp.support_load as sl
from supp.support_load import read_csv
from supp.support_constants import VIZIER_IDS, PATH_DATA, PATH_DESCRIPTION
from supp.support_save import save_df
from supp.support_analyzer import make_excel_analysis

In [2]:
# load database
dfs, dfs_name, dfs_export_date = sl.load_pickle()
iton, ntoi = sl.get_name_dicts(dfs_name)
dfs_export_date

Pickle database loaded.
C:\Users\Stoja\OneDrive\Documents\diplomka\scr\data\dfs_complete.pickle
Applied preprocessing: merge_on_jones
Applied preprocessing: remove_jones_duplicates


'2024-10-11'

In [3]:
# load dataframe with binary encoded titles
vizier__35_title_list = read_csv('vizier__35_title_list')
vizier__35_title_list = vizier__35_title_list.set_index('ID_person', drop=True)
print(vizier__35_title_list.shape)
vizier__35_title_list.head()

CSV file loaded.
C:\Users\Stoja\OneDrive\Documents\diplomka\scr\data\unpacked\df_person_all\merged\vizier__35_title_list.csv
(4962, 2183)


vizier  (cmcw ?) wxrt aAt  (wnn) imAx.c nfr xr Ra nb Axt(y)  \
ID_person                                                                
322             0                0.0                               0.0   
323             0                0.0                               0.0   
324             0                0.0                               0.0   
325             0                0.0                               0.0   
326             0                0.0                               0.0   

           ... 1p  ... BA-Nfr-ir-kA-ra  ... Hct n xnt[yw-S]  \
ID_person                                                     
322           0.0                  0.0                  0.0   
323           0.0                  0.0                  0.0   
324           0.0                  0.0                  0.0   
325           0.0                  0.0                  0.0   
326           0.0                  0.0                  0.0   

           ... Hm(w)-nTr ... m Cxt-Ra  ... Hm-kA  ... Hm-nTr 1wt-1r  \
ID_person                                                             
322                               0.0        0.0                0.0   
323                               0.0        0.0                0.0   
324                               0.0        0.0                0.0   
325                               0.0        0.0                0.0   
326                               0.0        0.0                0.0   

           ... Htpwt DfAw bity  ...  zmAwt mry Nbty ncwt  zwnw  zwnw Xt pr-aA  \
ID_person                       ...                                             
322                        0.0  ...                  0.0   0.0            0.0   
323                        0.0  ...                  0.0   0.0            0.0   
324                        0.0  ...                  0.0   0.0            0.0   
325                        0.0  ...                  0.0   0.0            0.0   
326                        0.0  ...                  0.0   0.0            0.0   

           zwnw cmcw pr-aA  zwnw irty  zwnw irty pr-aA  zwnw pr-aA  \
ID_person                                                            
322                    0.0        0.0              0.0         0.0   
323                    0.0        0.0              0.0         0.0   
324                    0.0        0.0              0.0         0.0   
325                    0.0        0.0              0.0         0.0   
326                    0.0        0.0              0.0         0.0   

           zš (n) (i)m(yw) zA 20 (n) niwt mr   \
ID_person                                       
322                                       0.0   
323                                       0.0   
324                                       0.0   
325                                       0.0   
326                                       0.0   

           zš (n) (i)m(yw) zA 20 (n) niwt mr Mn-anx-Nfr-kA-ra  \
ID_person                                                       
322                                                      0.0    
323                                                      0.0    
324                                                      0.0    
325                                                      0.0    
326                                                      0.0    

           zš wab 200 pr-aA  
ID_person                    
322                     0.0  
323                     0.0  
324                     0.0  
325                     0.0  
326                     0.0  

[5 rows x 2183 columns]

In [4]:
# read df sex of all persons
df_sex = read_csv('08_sex')
print(df_sex.shape)
df_sex.head()

CSV file loaded.
C:\Users\Stoja\OneDrive\Documents\diplomka\scr\data\unpacked\df_person_all\08_sex.csv
(4962, 3)


ID_person  female  male
0        322     0.0   1.0
1        323     0.0   1.0
2        324     0.0   1.0
3        325     0.0   1.0
4        326     0.0   1.0

### feature selection

In [5]:
# RESTRICTION
# select only viziers
df1 = vizier__35_title_list[vizier__35_title_list['vizier'] == 1]
#df1 = df1.drop(columns=['vizier'])
print(f'{df1.shape}\tshape of df1')
df1.head(2)

(80, 2183)	shape of df1


vizier  (cmcw ?) wxrt aAt  (wnn) imAx.c nfr xr Ra nb Axt(y)  \
ID_person                                                                
481             1                0.0                               0.0   
1               1                0.0                               0.0   

           ... 1p  ... BA-Nfr-ir-kA-ra  ... Hct n xnt[yw-S]  \
ID_person                                                     
481           0.0                  0.0                  0.0   
1             0.0                  0.0                  0.0   

           ... Hm(w)-nTr ... m Cxt-Ra  ... Hm-kA  ... Hm-nTr 1wt-1r  \
ID_person                                                             
481                               0.0        0.0                0.0   
1                                 0.0        0.0                0.0   

           ... Htpwt DfAw bity  ...  zmAwt mry Nbty ncwt  zwnw  zwnw Xt pr-aA  \
ID_person                       ...                                             
481                        0.0  ...                  0.0   0.0            0.0   
1                          0.0  ...                  0.0   0.0            0.0   

           zwnw cmcw pr-aA  zwnw irty  zwnw irty pr-aA  zwnw pr-aA  \
ID_person                                                            
481                    0.0        0.0              0.0         0.0   
1                      0.0        0.0              0.0         0.0   

           zš (n) (i)m(yw) zA 20 (n) niwt mr   \
ID_person                                       
481                                       0.0   
1                                         0.0   

           zš (n) (i)m(yw) zA 20 (n) niwt mr Mn-anx-Nfr-kA-ra  \
ID_person                                                       
481                                                      0.0    
1                                                        0.0    

           zš wab 200 pr-aA  
ID_person                    
481                     0.0  
1                       0.0  

[2 rows x 2183 columns]

In [6]:
# show persons with the fewest titles
df_help = df1[['vizier']].copy()
df_help['titel_total_count'] = df1.sum(axis=1)
df_help = df_help.sort_values('titel_total_count', ascending=True)
df_help.head(20)

vizier  titel_total_count
ID_person                           
481             1                2.0
3600            1                2.0
2698            1                2.0
3563            1                2.0
856             1                2.0
4490            1                4.0
3597            1                5.0
3601            1                5.0
56              1                6.0
3599            1                7.0
3593            1                8.0
216             1                8.0
3691            1                8.0
17              1                8.0
2578            1               10.0
199             1               10.0
3602            1               10.0
3608            1               12.0
3084            1               13.0
3011            1               13.0

In [7]:
# show persons with the most titles
df_help = df1[['vizier']].copy()
df_help['titel_total_count'] = df1.sum(axis=1)
df_help = df_help.sort_values('titel_total_count', ascending=False)
df_help.head(20)

vizier  titel_total_count
ID_person                           
3307            1              102.0
3697            1               65.0
3616            1               64.0
3612            1               62.0
1               1               59.0
1686            1               57.0
3229            1               56.0
3690            1               55.0
3326            1               54.0
3635            1               50.0
3689            1               47.0
3696            1               46.0
3415            1               45.0
1242            1               44.0
2254            1               43.0
2183            1               41.0
3227            1               39.0
70              1               35.0
124             1               34.0
3595            1               33.0

**NOTE:**
- There is about 15 viziers with relatively small number of title (up to 10).
- On the other side, there are viziers with more like 50 titles (one with 102).

In [8]:
# RESTRICTION
# take only titles that have accured more like 10 times
col_mask = (df1.sum(axis=0)>10)
df_vizier = df1.loc[:, df1.columns[col_mask]]
print(df_vizier.shape)
df_vizier.head()

(80, 46)


vizier  HAty-a  Hry-cStA  Hry-cStA n pr-dwAt  \
ID_person                                                 
481             1     0.0       0.0                 0.0   
1               1     1.0       0.0                 1.0   
56              1     0.0       0.0                 0.0   
856             1     0.0       0.0                 0.0   
1058            1     0.0       1.0                 0.0   

           Hry-cStA n wD(t)-mdw nbt nt ncwt  Hry-tp Nxb  Htc(?) Inpw  Xry-Hbt  \
ID_person                                                                       
481                                     0.0         0.0          0.0      0.0   
1                                       1.0         1.0          0.0      1.0   
56                                      0.0         0.0          0.0      1.0   
856                                     0.0         0.0          0.0      0.0   
1058                                    0.0         0.0          0.0      0.0   

           Xry-Hbt Hry-tp  Xry-tp ncwt  ...  tAyty TAty (n) zAb  \
ID_person                               ...                       
481                   0.0          0.0  ...                 1.0   
1                     1.0          0.0  ...                 1.0   
56                    0.0          0.0  ...                 1.0   
856                   0.0          0.0  ...                 1.0   
1058                  0.0          1.0  ...                 1.0   

           wr 5 (m) pr 9Hwty  xrp SnDt nbt  xrp aH  xrp iAwt nbwt nTrwt  \
ID_person                                                                 
481                      0.0           0.0     0.0                  0.0   
1                        0.0           0.0     1.0                  0.0   
56                       0.0           0.0     0.0                  0.0   
856                      0.0           0.0     0.0                  0.0   
1058                     0.0           0.0     0.0                  0.0   

           xtm(ty)-bity  zA ncwt  zA ncwt n Xt.f  zA ncwt n Xt.f cmcw  \
ID_person                                                               
481                 0.0      0.0             0.0                  0.0   
1                   0.0      1.0             1.0                  0.0   
56                  0.0      0.0             0.0                  0.0   
856                 0.0      0.0             0.0                  0.0   
1058                0.0      0.0             0.0                  0.0   

           zS mDAt-nTr  
ID_person               
481                0.0  
1                  1.0  
56                 0.0  
856                0.0  
1058               0.0  

[5 rows x 46 columns]

### correlation analysis

In [9]:
# Compute the correlation matrix
correlation_matrix = df_vizier.corr()
# Extract the upper triangle of the correlation matrix without the diagonal
mask = np.triu(np.ones(correlation_matrix.shape), k=1).astype(bool)
upper_triangle = correlation_matrix.where(mask)
# Flatten and sort the correlations by absolute value
correlation_pairs = upper_triangle.unstack()   #.dropna()
sorted_correlations = correlation_pairs.abs().sort_values(ascending=False)
# Display top correlations
df_correlation = sorted_correlations.reset_index()
df_correlation.columns = ['title_1', 'title_2', 'correlation']

print("Top Correlation Pairs:")
df_correlation.head()

Top Correlation Pairs:


title_1       title_2  correlation
0        mdw rxyt     iwn knmwt     0.898822
1    xrp SnDt nbt      cm / ctm     0.758600
2        mdw rxyt   Xry-tp ncwt     0.755929
3       iwn knmwt   Xry-tp ncwt     0.724314
4  imy-rA prwy-HD  imy-rA Snwty     0.689978

**NOTE:**
- The pair (`mdw rxyt`, `iwn knmwt`) has a correlation above 0.85.  
- For now, I will keep it in the dataset.  
- If it causes problems with multicollinearity, VIF will remove it.

### remove vizier titles

In [10]:
vizier_ids = VIZIER_IDS
vizier_ids

[293, 1830, 292]

In [11]:
# get names of vizier titles
df_titles = dfs[ntoi['df_titles']]
vizier_title_name = df_titles.loc[df_titles['ID_title'].isin(vizier_ids), 'title'].to_list()
vizier_title_name

['tAyty TAty (n) zAb', 'tAyty TAty (n) zAb mAa']

In [12]:
columns_all = df_vizier.columns
non_veziers_titles = list(set(columns_all) - set(vizier_title_name))
df_features = df_vizier[non_veziers_titles].copy()
print(f'{df_vizier.shape}\tshape of df with all features')
print(f'{df_features.shape}\tshape of df with non-vizier features')
df_features.head()

(80, 46)	shape of df with all features
(80, 45)	shape of df with non-vizier features


imy iz  imy-rA Snwty  imy-rA wabty  cm / ctm  aD-mr (n) zAb  \
ID_person                                                                
481           0.0           0.0           0.0       0.0            0.0   
1             0.0           0.0           0.0       0.0            0.0   
56            0.0           0.0           0.0       0.0            0.0   
856           0.0           0.0           0.0       0.0            0.0   
1058          0.0           0.0           0.0       0.0            1.0   

           imy-rA Hwt-wrt 6  xrp SnDt nbt  zA ncwt  imy-rA gc-pr  iry-pat  \
ID_person                                                                   
481                     0.0           0.0      0.0           0.0      0.0   
1                       0.0           0.0      1.0           0.0      0.0   
56                      0.0           0.0      0.0           0.0      0.0   
856                     0.0           0.0      0.0           0.0      0.0   
1058                    1.0           0.0      0.0           0.0      0.0   

           ...  Hry-cStA  vizier  imy-rA prwy-HD  imy-rA prwy-nbw  HAty-a  \
ID_person  ...                                                              
481        ...       0.0       1             0.0              0.0     0.0   
1          ...       0.0       1             1.0              1.0     1.0   
56         ...       0.0       1             0.0              0.0     0.0   
856        ...       0.0       1             0.0              0.0     0.0   
1058       ...       1.0       1             0.0              0.0     0.0   

           iwn knmwt  Xry-Hbt  imy-rA zS(w) a(w) (nw) ncwt  mdw rxyt  \
ID_person                                                              
481              0.0      0.0                          0.0       0.0   
1                0.0      1.0                          0.0       0.0   
56               0.0      1.0                          0.0       0.0   
856              0.0      0.0                          0.0       0.0   
1058             1.0      0.0                          0.0       1.0   

           zA ncwt n Xt.f cmcw  
ID_person                       
481                        0.0  
1                          0.0  
56                         0.0  
856                        0.0  
1058                       0.0  

[5 rows x 45 columns]

In [13]:
# Control that vizier titles were removed
# MUST BE FALSE
any(vizier_title in df_features.columns for vizier_title in vizier_title_name)

False

### make final dataframe

In [14]:
# define list with most common titles among viziers 
top_vizier_titles = df_features.columns.to_list()
top_vizier_titles = list(set(top_vizier_titles) - set(['ID_person', 'vizier']))
print(f'{len(top_vizier_titles)}\ttitles selected as the final selection')

44	titles selected as the final selection


In [15]:
# take only selected features into final dataframe
columns_to_keep = ['vizier'] + top_vizier_titles
df_result = vizier__35_title_list[columns_to_keep]
print(df_result.shape)
df_result.head()

(4962, 45)


vizier  imy iz  imy-rA Snwty  imy-rA wabty  cm / ctm  \
ID_person                                                         
322             0     0.0           0.0           0.0       0.0   
323             0     0.0           0.0           0.0       0.0   
324             0     0.0           0.0           0.0       0.0   
325             0     0.0           0.0           0.0       0.0   
326             0     0.0           0.0           0.0       0.0   

           aD-mr (n) zAb  imy-rA Hwt-wrt 6  xrp SnDt nbt  zA ncwt  \
ID_person                                                           
322                  0.0               0.0           0.0      0.0   
323                  0.0               0.0           0.0      0.0   
324                  0.0               0.0           0.0      0.0   
325                  0.0               0.0           0.0      0.0   
326                  0.0               0.0           0.0      0.0   

           imy-rA gc-pr  ...  xtm(ty)-bity  Hry-cStA  imy-rA prwy-HD  \
ID_person                ...                                           
322                 0.0  ...           0.0       0.0             0.0   
323                 0.0  ...           0.0       0.0             0.0   
324                 0.0  ...           0.0       0.0             0.0   
325                 0.0  ...           0.0       0.0             0.0   
326                 0.0  ...           0.0       0.0             0.0   

           imy-rA prwy-nbw  HAty-a  iwn knmwt  Xry-Hbt  \
ID_person                                                
322                    0.0     0.0        0.0      0.0   
323                    0.0     0.0        0.0      0.0   
324                    0.0     0.0        0.0      0.0   
325                    0.0     0.0        0.0      0.0   
326                    0.0     0.0        0.0      0.0   

           imy-rA zS(w) a(w) (nw) ncwt  mdw rxyt  zA ncwt n Xt.f cmcw  
ID_person                                                              
322                                0.0       0.0                  0.0  
323                                0.0       0.0                  0.0  
324                                0.0       0.0                  0.0  
325                                0.0       0.0                  0.0  
326                                0.0       0.0                  0.0  

[5 rows x 45 columns]

### restriction to only male persons

**NOTE:**
- Only males could become viziers.  
- Therefore, I restrict the dataset to males only.

In [16]:
print(f'{df_sex.shape}\tshape of df_sex')
df_sex.head()

(4962, 3)	shape of df_sex


ID_person  female  male
0        322     0.0   1.0
1        323     0.0   1.0
2        324     0.0   1.0
3        325     0.0   1.0
4        326     0.0   1.0

In [17]:
id_male = df_sex.loc[df_sex['male'] == 1, 'ID_person'].to_list()
print(f'{len(id_male)}\tcount of males')

3930	count of males


In [18]:
# RESTRICTION
# take only males
df_result_male = df_result.loc[df_result.index.isin(id_male), :]
print(df_result_male.shape)
df_result_male.head()

(3930, 45)


vizier  imy iz  imy-rA Snwty  imy-rA wabty  cm / ctm  \
ID_person                                                         
322             0     0.0           0.0           0.0       0.0   
323             0     0.0           0.0           0.0       0.0   
324             0     0.0           0.0           0.0       0.0   
325             0     0.0           0.0           0.0       0.0   
326             0     0.0           0.0           0.0       0.0   

           aD-mr (n) zAb  imy-rA Hwt-wrt 6  xrp SnDt nbt  zA ncwt  \
ID_person                                                           
322                  0.0               0.0           0.0      0.0   
323                  0.0               0.0           0.0      0.0   
324                  0.0               0.0           0.0      0.0   
325                  0.0               0.0           0.0      0.0   
326                  0.0               0.0           0.0      0.0   

           imy-rA gc-pr  ...  xtm(ty)-bity  Hry-cStA  imy-rA prwy-HD  \
ID_person                ...                                           
322                 0.0  ...           0.0       0.0             0.0   
323                 0.0  ...           0.0       0.0             0.0   
324                 0.0  ...           0.0       0.0             0.0   
325                 0.0  ...           0.0       0.0             0.0   
326                 0.0  ...           0.0       0.0             0.0   

           imy-rA prwy-nbw  HAty-a  iwn knmwt  Xry-Hbt  \
ID_person                                                
322                    0.0     0.0        0.0      0.0   
323                    0.0     0.0        0.0      0.0   
324                    0.0     0.0        0.0      0.0   
325                    0.0     0.0        0.0      0.0   
326                    0.0     0.0        0.0      0.0   

           imy-rA zS(w) a(w) (nw) ncwt  mdw rxyt  zA ncwt n Xt.f cmcw  
ID_person                                                              
322                                0.0       0.0                  0.0  
323                                0.0       0.0                  0.0  
324                                0.0       0.0                  0.0  
325                                0.0       0.0                  0.0  
326                                0.0       0.0                  0.0  

[5 rows x 45 columns]

### save final dataset

In [19]:
save_df(df_result_male, 'df_vizier_titles_v1', save_index=True)
print(f'{df_result_male.shape}\tshape of df_result_male')
df_result_male.head()

Dataframe saved into C:\Users\Stoja\OneDrive\Documents\diplomka\scr\data\df_vizier_titles_v1.csv
(3930, 45)	shape of df_result_male


vizier  imy iz  imy-rA Snwty  imy-rA wabty  cm / ctm  \
ID_person                                                         
322             0     0.0           0.0           0.0       0.0   
323             0     0.0           0.0           0.0       0.0   
324             0     0.0           0.0           0.0       0.0   
325             0     0.0           0.0           0.0       0.0   
326             0     0.0           0.0           0.0       0.0   

           aD-mr (n) zAb  imy-rA Hwt-wrt 6  xrp SnDt nbt  zA ncwt  \
ID_person                                                           
322                  0.0               0.0           0.0      0.0   
323                  0.0               0.0           0.0      0.0   
324                  0.0               0.0           0.0      0.0   
325                  0.0               0.0           0.0      0.0   
326                  0.0               0.0           0.0      0.0   

           imy-rA gc-pr  ...  xtm(ty)-bity  Hry-cStA  imy-rA prwy-HD  \
ID_person                ...                                           
322                 0.0  ...           0.0       0.0             0.0   
323                 0.0  ...           0.0       0.0             0.0   
324                 0.0  ...           0.0       0.0             0.0   
325                 0.0  ...           0.0       0.0             0.0   
326                 0.0  ...           0.0       0.0             0.0   

           imy-rA prwy-nbw  HAty-a  iwn knmwt  Xry-Hbt  \
ID_person                                                
322                    0.0     0.0        0.0      0.0   
323                    0.0     0.0        0.0      0.0   
324                    0.0     0.0        0.0      0.0   
325                    0.0     0.0        0.0      0.0   
326                    0.0     0.0        0.0      0.0   

           imy-rA zS(w) a(w) (nw) ncwt  mdw rxyt  zA ncwt n Xt.f cmcw  
ID_person                                                              
322                                0.0       0.0                  0.0  
323                                0.0       0.0                  0.0  
324                                0.0       0.0                  0.0  
325                                0.0       0.0                  0.0  
326                                0.0       0.0                  0.0  

[5 rows x 45 columns]